### PR TITLE
Adding RankTrainValidationSplitter

### PR DESCRIPTION
**Pull request by dciborow**
Adding RankTrainValidationSplitter

### CODE REVIEW

**Comment by dciborow** [comment]
@mhamilton723 this is ready for review.

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+      case n =>
+        ExecutionContext.fromExecutorService(SparkHelpers.getThreadUtils()
+          .newDaemonCachedThreadPool(s"${this.getClass.getSimpleName}-thread-pool", n))
+    }
+  }
+
+  override def fit(dataset: Dataset[_]): RankingTrainValidationSplitModel = {
+    val schema = dataset.schema
+    transformSchema(schema, logging = true)
+    val est = $(estimator)
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: nit: change $ getters to use your getters

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+    val executionContext = getExecutionContext
+
+    def calculateMetrics(model: Transformer, validationDataset: Dataset[_]): Double = model match {
+      case p: PipelineModel =>
+        //Assume Rec Algo is last stage of pipeline
+        val modelTemp = model.asInstanceOf[PipelineModel].stages.last
+        calculateMetrics(modelTemp, validationDataset)
+      case a: ALSModel =>
+        val recs = model.asInstanceOf[ALSModel].recommendForAllUsers(5 * eval.getK)
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: magic number?

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    }
+
+    val metricFutures = epm.zipWithIndex.map { case (paramMap, paramIndex) =>
+      Future[Double] {
+        val model = est.fit(trainingDataset, paramMap).asInstanceOf[Model[_]]
+        calculateMetrics(model, validationDataset)
+      }(executionContext)
+    }
+
+    val metrics = metricFutures.map(SparkHelpers.getThreadUtils().awaitResult(_, Duration.Inf))
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: does spark use an infinite timeout for this?

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+      .withColumnRenamed("count(" + $(itemCol) + ")", "nitems")
+      .where(col("nitems") >= $(minRatingsU))
+      .drop("nitems")
+      .cache()
+  }
+
+  private def filterByUserRatingCount(dataset: Dataset[_]): DataFrame = dataset
+    .groupBy($(itemCol))
+    .agg(col($(itemCol)), count(col($(userCol))))
+    .withColumnRenamed("count(" + $(userCol) + ")", "ncustomers")
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: nit: can use string interpolation s"count(${getUserCol})"

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+        .groupBy(col($(userCol)))
+        .agg(collect_list(col("itemIDRating")))
+        .withColumn("shuffle", shuffle(col("collect_list(itemIDRating)")))
+        .withColumn("train", sliceudf(col("shuffle")))
+        .withColumn("test", dropudf(col("shuffle")))
+        .drop(col("collect_list(itemIDRating)")).drop(col("shuffle"))
+        .cache()
+
+      val popLeft = udf((r: mutable.WrappedArray[Double]) => r(0))
+      val popRight = udf((r: mutable.WrappedArray[Double]) => r(1))
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: can just use col("foo").getItem(0) or 1. This way spark will support query pushdown through it

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    val joined_rec_actual = perUserRecommendedItemsDF
+      .join(perUserActualItemsDF, userColumn)
+      .drop(userColumn)
+
+    joined_rec_actual
+  }
+}
+
+object RankingTrainValidationSplit extends ComplexParamsReadable[RankingTrainValidationSplit] {
+  def popRow(r: Row): Any = r.getDouble(1)
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: maybe call this method something a bit more interpretable, like getID or w/e

File: src/recommendation/src/test/scala/RankingTrainValidationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      .setTrainRatio(0.8)
+      .setUserCol(customerIndex.getOutputCol)
+      .setItemCol(itemIndex.getOutputCol)
+      .setRatingCol("rating")
+
+    val tvModel = tvRecommendationSplit.fit(transformedDf)
+
+    val model = tvModel.getBestModel.asInstanceOf[ALSModel]
+
+    val items = model.recommendForAllUsers(3)
```
Comment (mhamilton723) [inline src/recommendation/src/test/scala/RankingTrainValidationSpec.scala]: is this lazy or will this print/return always

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala (hunk 3 above)
Comment (dciborow) [reply src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: image

Comment (dciborow) [reply src/recommendation/src/main/scala/RankingTrainValidationSplit.scala]: yup.

File: src/recommendation/src/test/scala/RankingTrainValidationSpec.scala (hunk 7 above)
Comment (dciborow) [reply src/recommendation/src/test/scala/RankingTrainValidationSpec.scala]: this is not lazy

Comment (mhamilton723) [reply src/recommendation/src/test/scala/RankingTrainValidationSpec.scala]: If it returns a dataframe then we should test that this dataframe should have the right structure

**Pull request merged**